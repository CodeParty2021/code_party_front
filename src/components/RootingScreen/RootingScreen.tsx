import { SignOutButton } from "components/SignOutButton/SignOutButton";
import React from "react";
import { useSelector } from "react-redux";
import { Link } from "react-router-dom";
import { RootState } from "store";

type Props = {};

export const RootingScreen: React.FC<Props> = () => {
  const { user, isLogin } = useSelector((state: RootState) => state.user);
  return (
    <div>
      <p> ==============</p>
      <h3>以下開発用画面</h3>

      <a href="https://strap.app/app/board/jkJsgBr4NKM8IvxIYo8s">
        画面名称はこちら
      </a>
      <ul>
        <li>
          <Link to="/">トップページ</Link>
        </li>
        <li>
          <Link to="/stages">ステージ選択画面</Link>
        </li>
        <li>
          <Link to="/start">ログイン画面</Link>
        </li>
        <li>
          <Link to="/mode-select">モード選択画面</Link>
        </li>
        <li>
          <Link to="/casual-battle">カジュアルロビー画面</Link>
        </li>
        <li>
          <Link to="/casual-battle/waiting-room">待機招待画面</Link>
        </li>
        <li>
          <Link to="/casual-battle/search-room">ルーム参加画面</Link>
        </li>
        <li>
          <Link to="/casual-battle/result">観戦画面</Link>
        </li>
        <li>
          <Link to="/free-coding">フリーコーディング</Link>
        </li>
        <li>
          <Link to="/codes">ガレージ画面（コード一覧)</Link>
        </li>
      </ul>
      <h1>ログインユーザ</h1>
      <p>{user?.displayName}さん</p>
      {isLogin && <SignOutButton />}
    </div>
  );
};
