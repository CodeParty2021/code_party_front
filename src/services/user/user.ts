import { createSlice } from "@reduxjs/toolkit";
import firebase from "firebase/compat/app";
import axios, { AxiosResponse } from "axios";
import { uri } from "config";
import { onAuthStateChanged } from "firebase/auth";
import { auth } from "firebase_config";
export type User = {
  id: string;
  displayName: string;
  email: string;
  picture: string;
  jwt: string; //これをheaderに入れてAPIやり取りする。
};

export type UserState = {
  user: User | null;
  isLogin: boolean;
  unRegisterObserver: firebase.Unsubscribe | null; // ログイン監視用のobserverのkill用メソッド
};

type UserAuthResponse = {
  // レスポンスの型
  userInfo: {
    displayName: string;
    email: string;
    id: string;
    picture: string;
  };
  isCreated: boolean;
};

type signInAction = { type: string; payload: User };
type setUnRegisterObserverAction = {
  type: string;
  payload: firebase.Unsubscribe;
};

const isUserAuthResponse = (data: any): data is UserAuthResponse => {
  return data !== undefined;
};
// createSlice() で actions と reducers を一気に生成
// https://scrapbox.io/frontend-akihito/Firebase_Auth_%E3%81%AEonAuthStateObserver%E3%81%AB%E3%81%A4%E3%81%84%E3%81%A6

const userSlice = createSlice({
  name: "users", //識別用の名前
  initialState: { user: null, isLogin: false } as UserState,
  reducers: {
    signIn: (state, action: signInAction) => {
      state.isLogin = true;
      state.user = action.payload;
    },
    setUnRegisterObserver: (state, action: setUnRegisterObserverAction) => {
      state.unRegisterObserver = action.payload;
    },
    signOut: (state) => {
      state.isLogin = false;
      state.user = null;
    },
  },
});

// 今回追加したgetRequestもエクスポートする
export const { signIn, signOut, setUnRegisterObserver } = userSlice.actions;

// RESTAPIの実行とgetRequestの呼び出しをする
export const signInAsync = () => {
  //dispatch関数を返す
  return async (dispatch: any) => {
    // unRegisterObserverメソッドはObserverを捨てる関数
    const unRegisterObserver = firebase.auth().onAuthStateChanged((user) => {
      if (user) {
        user.getIdToken().then((idToken: string) => {
          // promiseが戻り値のときはこういう書き方をする
          axios
            .get(uri + "/users/auth", {
              headers: {
                "Content-Type": "multipart/form-data",
                Authorization: "Bearer " + idToken,
              },
            })
            .then((res: AxiosResponse<UserAuthResponse>) => {
              if (isUserAuthResponse(res.data)) {
                const userInfo = res.data.userInfo;
                dispatch(
                  signIn({
                    id: userInfo.id,
                    displayName: userInfo.displayName,
                    email: userInfo.email as string,
                    picture: userInfo.picture,
                    jwt: idToken,
                  })
                );
                if (res.data.isCreated) {
                  console.log("サインイン");
                  console.log(idToken);
                  //TODO: サインイン時の処理を書く
                } else {
                  console.log("ログイン");
                  console.log(idToken);
                  //TODO: ログイン時の処理を書く
                }
              }
            })
            .catch();
        });
      }
    });
    dispatch(setUnRegisterObserver(unRegisterObserver));
  };
};

export const signOutAsync = () => {
  return async (dispatch: any) => {
    firebase
      .auth()
      .signOut()
      .then(() => {
        dispatch(signOut());
      });
  };
};

export const setCallBackToSyncUser = () => {
  return async (dispatch: any) => {
    onAuthStateChanged(auth, (user) => {
      if (user) {
        dispatch(signInAsync());
      }
    });
  };
};

export default userSlice.reducer;
