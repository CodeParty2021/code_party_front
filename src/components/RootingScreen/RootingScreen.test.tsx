import React from "react";
import { shallow } from "enzyme";
import { RootingScreen } from "./RootingScreen";
import { useDispatch, useSelector } from "react-redux";
import { signOut, LoginUserState } from "../../services/user/user";

jest.mock("react-redux");

const useSelectorMock = useSelector as jest.Mock<LoginUserState>;
const useDispatchMock = useDispatch as jest.Mock;

describe("<RootingScreen />", () => {
  beforeEach(() => {
    useSelectorMock.mockReturnValue({
      user: null,
      isLogin: false,
      unRegisterObserver: null,
      loading: false,
    });
    useDispatchMock.mockReturnValue(jest.fn());
  });
  afterEach(() => {
    jest.resetAllMocks();
  });
  it("RootingScreen snapshot test", () => {
    useDispatchMock.mockReturnValue(signOut());
    const wrapper = shallow(<RootingScreen />);
    expect(wrapper.getElements()).toMatchSnapshot();
  });
});
