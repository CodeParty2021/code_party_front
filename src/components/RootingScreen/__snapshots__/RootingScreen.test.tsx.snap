// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RootingScreen /> RootingScreen snapshot test 1`] = `
Array [
  <div
    style={
      Object {
        "position": "absolute",
        "top": "100%",
      }
    }
  >
    <p>
       ==============
    </p>
    <h3>
      以下開発用画面
    </h3>
    <a
      href="https://strap.app/app/board/jkJsgBr4NKM8IvxIYo8s"
    >
      画面名称はこちら
    </a>
    <ul>
      <li>
        <Link
          to="/"
        >
          トップページ
        </Link>
      </li>
      <li>
        <Link
          to="/stages"
        >
          ステージ選択画面
        </Link>
      </li>
      <li>
        <Link
          to="/start"
        >
          ログイン画面
        </Link>
      </li>
      <li>
        <Link
          to="/select-mode"
        >
          選択画面Top
        </Link>
      </li>
      <li>
        <Link
          to="/free-coding"
        >
          フリーコーディング
        </Link>
      </li>
      <li>
        <Link
          to="/codes"
        >
          コード一覧
        </Link>
      </li>
      <li>
        <Link
          to="/Lp"
        >
          LP画面
        </Link>
      </li>
    </ul>
    <h1>
      ロボット開発モード系のページ
    </h1>
    <ul>
      <li>
        <Link
          to="/robot-development/top"
        >
          ロボット開発モードの選択画面
        </Link>
      </li>
      <li>
        <Link
          to="/robot-development/specification"
        >
          ロボットの設計仕様書
        </Link>
      </li>
      <li>
        <Link
          to="/garage"
        >
          ロボットのステータスを見る
        </Link>
      </li>
    </ul>
    <h1>
      ルーム対戦モード系のページ
    </h1>
    <ul>
      <li>
        <Link
          to="/room-match"
        >
          カジュアルロビー画面
        </Link>
      </li>
      <li>
        <Link
          to="/room-match/invitation/:roomId"
        >
          カジュアル対戦招待画面
        </Link>
      </li>
      <li>
        <Link
          to="/room-match/waiting-room"
        >
          待機招待画面
        </Link>
      </li>
      <li>
        <Link
          to="/room-match/result"
        >
          観戦画面
        </Link>
      </li>
      <li>
        <Link
          to="/room-match/pick-code"
        >
          対戦させるロボットを選ぶ
        </Link>
      </li>
    </ul>
    <h1>
      チュートリアル系のページ
    </h1>
    <ul>
      <li>
        <Link
          to="/tutorial/world/:id/missions"
        >
          ミッション一覧
        </Link>
      </li>
      <li>
        <Link
          to="/tutorial/world/:world_id/mission/:mission_id/step/:step_id"
        >
          ミッションの詳細(step)用のページ
        </Link>
      </li>
      <li>
        <Link
          to="/tutorial/world/:world_id/mission/:mission_id/end"
        >
          ミッション終了後のページ
        </Link>
      </li>
      <li>
        <Link
          to="/tutorial/world/:world_id/intro"
        >
          ワールド・ゲームの説明画面
        </Link>
      </li>
    </ul>
    <h1>
      イベント系のページ
    </h1>
    <ul>
      <li>
        <Link
          to="/event"
        >
          トップページ
        </Link>
      </li>
      <li>
        <Link
          to="/event/select-mode"
        >
          モード選択
        </Link>
      </li>
      <li>
        <Link
          to="/event/set-name"
        >
          名前設定
        </Link>
      </li>
      <li>
        <Link
          to="/event/select-ai"
        >
          対戦相手の選択
        </Link>
      </li>
    </ul>
    <h1>
      ログインユーザ
    </h1>
    <p>
      さん
    </p>
  </div>,
]
`;
